### PR TITLE
cmd/charm: refactor resource upload before adding docker resources

### PR DESCRIPTION
**Pull request by mhilton**

### CODE REVIEW

File: cmd/charm/charmcmd/attach.go [hunk 1]
```diff
@@ -93,12 +93,20 @@ func (c *attachCommand) Run(ctxt *cmd.Context) error {
...
+	}
+
 	rev, err := uploadResource(uploadResourceParams{
 		ctxt:         ctxt,
 		client:       client,
 		charmId:      c.id,
+		meta:         &meta.CharmMetadata,
 		resourceName: c.resourceName,
-		filePath:     c.filePath,
+		reference:    c.filePath,
```
Comment (rogpeppe) [inline cmd/charm/charmcmd/attach.go]: Please change c.filePath to c.reference.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore-client/205/

File: cmd/charm/charmcmd/push.go [hunk 2]
```diff
@@ -203,9 +202,10 @@ func (c *pushCommand) pushResources(ctxt *cmd.Context, client *csClient, meta *c
 		if err := c.uploadResource(uploadResourceParams{
 			ctxt:         ctxt,
 			client:       client,
+			meta:         meta,
 			charmId:      c.id,
 			resourceName: resourceName,
-			filePath:     filePath,
+			reference:    filePath,
```
Comment (rogpeppe) [inline cmd/charm/charmcmd/push.go]: The above AbsPath call doesn't look quite right to me - we don't know it's a resource of file type yet.

**Review by rogpeppe — APPROVED** [review]
LGTM modulo a couple of suggestions, thanks.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore-client/206/

File: cmd/charm/charmcmd/attach.go [hunk 3]
```diff
@@ -119,18 +119,18 @@ func (c *attachCommand) Run(ctxt *cmd.Context) error {
 
 // parseResourceFileArg converts the provided string into a name and
 // filename. The string must be in the "<name>=<filename>" format.
-func parseResourceFileArg(raw string) (name string, filename string, err error) {
+func parseResourceFileArg(raw string) (name string, reference string, err error) {
```
Comment (rogpeppe) [inline cmd/charm/charmcmd/attach.go]: This function name and the doc comment seems a bit out of date now.

**Review by rogpeppe — APPROVED** [review]
LGTM with one final suggestion.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore-client/207/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-gce.jujugui.org:8080//job/charmstore-client/208/

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/charmstore-client-merge

**Pull request merged**